File: app/src/main/java/com/lessonscontrol/data/repositories/StudentRepository.java
package com.lessonscontrol.data.repositories;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.os.AsyncTask;

import com.lessonscontrol.data.AppRoomDatabase;
import com.lessonscontrol.data.dao.StudentDAO;
import com.lessonscontrol.data.entities.Student;

import java.util.List;

/**
 * {@link Student} repository.
 *
 * @author hblonski
 */
public class StudentRepository {

    private StudentDAO studentDAO;

    private LiveData<List<Student>> allStudents;

    public StudentRepository(Application application) {
        AppRoomDatabase appRoomDatabase = AppRoomDatabase.getInstance(application);
        studentDAO = appRoomDatabase.studentDAO();
        allStudents = studentDAO.getAllStudents();
    }

    public LiveData<List<Student>> getAllStudents() {
        return this.allStudents;
    }

    public void insert(Student student) {
        new InsertAsyncTask(studentDAO).execute(student);
    }

    public void delete(Student student) {
        new DeleteAsyncTask(studentDAO).execute(student);
    }

    public void update(Student student) {
        new UpdateAsyncTask(studentDAO).execute(student);
    }

    private static class InsertAsyncTask extends AsyncTask<Student, Void, Void> {

        private StudentDAO studentDAO;

        InsertAsyncTask(StudentDAO studentDAO) {
            this.studentDAO = studentDAO;
        }

        @Override
        protected Void doInBackground(final Student... students) {
            studentDAO.insert(students[0]);
            return null;
        }
    }

    private static class UpdateAsyncTask extends AsyncTask<Student, Void, Void> {

        private StudentDAO studentDAO;

        UpdateAsyncTask(StudentDAO studentDAO) {
            this.studentDAO = studentDAO;
        }

        @Override
        protected Void doInBackground(final Student... students) {
            studentDAO.update(students[0]);
            return null;
        }
    }

    private static class DeleteAsyncTask extends AsyncTask<Student, Void, Void> {

        private StudentDAO studentDAO;

        DeleteAsyncTask(StudentDAO studentDAO) {
            this.studentDAO = studentDAO;
        }

        @Override
        protected Void doInBackground(final Student... students) {
            studentDAO.delete(students[0]);
            return null;
        }
    }
}

File: app/src/main/java/com/lessonscontrol/data/repositories/StudentRepository.kt
package com.lessonscontrol.data.repositories

import android.app.Application
import android.arch.lifecycle.LiveData
import android.os.AsyncTask

import com.lessonscontrol.data.AppRoomDatabase
import com.lessonscontrol.data.dao.StudentDAO
import com.lessonscontrol.data.entities.Student

/**
 * [Student] repository.
 *
 * @author hblonski
 */
class StudentRepository(application: Application) {

    private val studentDAO: StudentDAO

    val allStudents: LiveData<List<Student>>

    init {
        val appRoomDatabase = AppRoomDatabase.getInstance(application)
        studentDAO = appRoomDatabase!!.studentDAO()
        allStudents = studentDAO.allStudents
    }

    fun insert(student: Student) {
        InsertAsyncTask(studentDAO).execute(student)
    }

    fun delete(student: Student) {
        DeleteAsyncTask(studentDAO).execute(student)
    }

    fun update(student: Student) {
        UpdateAsyncTask(studentDAO).execute(student)
    }

    private class InsertAsyncTask internal constructor(private val studentDAO: StudentDAO) : AsyncTask<Student, Void, Void>() {

        override fun doInBackground(vararg students: Student): Void? {
            studentDAO.insert(students[0])
            return null
        }
    }

    private class UpdateAsyncTask internal constructor(private val studentDAO: StudentDAO) : AsyncTask<Student, Void, Void>() {

        override fun doInBackground(vararg students: Student): Void? {
            studentDAO.update(students[0])
            return null
        }
    }

    private class DeleteAsyncTask internal constructor(private val studentDAO: StudentDAO) : AsyncTask<Student, Void, Void>() {

        override fun doInBackground(vararg students: Student): Void? {
            studentDAO.delete(students[0])
            return null
        }
    }
}
